File: matlab/ex5/learningCurve.m
function [error_train,error_val]=learningCurve(X,y,Xval,yval,lambda)
%% 函数功能：计算不同训练样本下的优化结果在训练集和验证集上的误差曲线
m = size(X,1);
error_train = zeros(m,1);
error_val = zeros(m,1);
% 每次带入不同数目的训练样本数量，计算在所带入的训练集上的代价，以及在验证集上的代价
for i = 1:m
    theta = trainLinearReg(X(1:i,:),y(1:i),lambda);
    [error_train(i),~] = linearRegCostFunction(X(1:i , :),y(1:i),theta,lambda);
    [error_val(i),~] = linearRegCostFunction(Xval,yval,theta,lambda);
end

end

File: python/ex5/learningCurve.py
import numpy as np

from linearCostFunction import linear_cost_function
from trainLinearRegression import train_linear_reg

def learning_curve(X,Y,Xval,Yval,lmd):
	m = X.shape[0]
	error_train = np.zeros(m)
	error_val = np.zeros(m)

	for num in range(m):
		theta = train_linear_reg(X[0:num+1,:],Y[0:num+1],lmd)

		error_train[num],_ = linear_cost_function(X[0:num+1,:],Y[0:num+1],theta,lmd)
		error_val[num],_ = linear_cost_function(Xval,Yval,theta,lmd)

	return error_train,error_val